### PR TITLE
Destination Redshift: Limit Standard insert statement size < 16MB

### PR DESCRIPTION
**Pull request by gisripa**
Limit the multi-value insert statement to not reach beyond [16MB](https://docs.aws.amazon.com/redshift/latest/dg/c_redshift-sql.html)

### CODE REVIEW

**Comment by gisripa** [comment]
* **#36973** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36973?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @gisripa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftInsertDestination.java [hunk 1]
```diff
@@ -45,12 +45,19 @@ public class RedshiftInsertDestination extends AbstractJdbcDestination<RedshiftS
       JdbcUtils.SSL_KEY, "true",
       "sslfactory", "com.amazon.redshift.ssl.NonValidatingFactory");
 
+  // insert into stmt has ~200 bytes
+  // Per record overhead of ~150 bytes for strings in statement like JSON_PARSE.. uuid etc
+  // If the flush size allows the max batch of 10k records, then net overhead is ~1.5MB.
+  // Lets round it to 2MB and keep a max buffer of 14MB
+  // This will allow not sending record set larger than 14M limiting the batch insert statement.
+  private static final Long MAX_BATCH_SIZE_FOR_FLUSH = 14 * 1024 * 1024L;
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftInsertDestination.java:53]: nit: how is the optimalBatchSize parameter used? I.e. does the async framework treat it as a recommendation, or a hard limit? (and let's align the naming to either optimalBatchSize + OPTIMAL_BATCH_SIZE_FOR_FLUSH, or maxBatchSize + MAX_BATCH_SIZE_FOR_FLUSH)

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java [hunk 2]
```diff
@@ -16,4 +27,49 @@ protected ObjectNode getBaseConfig() {
     return (ObjectNode) Jsons.deserialize(IOs.readFile(Path.of("secrets/1s1t_config.json")));
   }
 
+  @Test
+  public void testStandardInsertBatchSizeGtThan16Mb() throws Exception {
```
Comment (edgao) [inline airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java:31]: should we actually move this into the base TD test class? Seems useful for all destinations to verify handling of nontrivial data size

**Review by edgao — APPROVED** [review]
for my understanding: this is basically just telling the async framework to send us batches of <14MB, which hopefully gives us enough wiggle room to avoid generating sql > 16MB?

(I also have no idea how the async framework measures record size, is it using the serialized json string size?)

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java (hunk 2 above)
Comment (gisripa) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java:31]: Do we have limitations in staging batch size u recall. put it in std insert flow for SQL stmt limit. probably can pull it up to AbstractRedshiftTD too.

**Comment by gisripa** [comment]
> for my understanding: this is basically just telling the async framework to send us batches of <14MB, which hopefully gives us enough wiggle room to avoid generating sql > 16MB?
> 
> (I also have no idea how the async framework measures record size, is it using the serialized json string size?)

Yeah your understanding is correct.  Its using raw string size which is consumed from Std.in passed along into as record size.   If a bunch of small records add up to 14M, then our 10k batch insert will kick-in, else this will act as the guardrail and our batch will be << 10k

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/JdbcBufferedConsumerFactory.kt [hunk 3]
```diff
@@ -52,7 +52,8 @@ import org.slf4j.LoggerFactory
 object JdbcBufferedConsumerFactory {
     private val LOGGER: Logger = LoggerFactory.getLogger(JdbcBufferedConsumerFactory::class.java)
 
-    @JvmOverloads
```
Comment (gisripa) [inline airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/jdbc/JdbcBufferedConsumerFactory.kt:55]: No longer used in any Java context, so removed the Overloads.

File: airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java (hunk 2 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-redshift/src/test-integration/java/io/airbyte/integrations/destination/redshift/typing_deduping/RedshiftStandardInsertsTypingDedupingTest.java:31]: :shrug:

we probably target a specific size, but I'm not aware of redshift (or any staging dest) setting a hard limit on file size (and we're probably not reaching that limit anyway...)

**Comment by stephane-airbyte** [comment]
@gisripa can we create an issue to solve this in a more holistic way? I really don't think this should be solved at the individual connector level. We need to start thinking of a way for various connectors to advertise their capabilities to the platform rather than pass more and more parameters to the CDK.

**Comment by edgao** [comment]
+1 on solving this generically, but I do think this needs to be solved in-connector (and therefore probably in the cdk). It's a limitation of the length of the sql statement `insert into <table> (cols...) values (rec1...), (rec2...), ...` - platform has no idea that we're converting record messages into a giant sql blob

But right now there's a lot of indirection from stdin -> async framework (?) -> jdbc sql operations (I think??) -> redshift sql operations (???) -> actual insert statement, which makes this kind of painful :(

**Pull request merged**